### PR TITLE
[21] Add API login

### LINKED ISSUE #21: As a user, I can make an API call to sign in to my account
## Why

We need to encode an access token that we can use to represent the session. This token should time out after a period of time.

## Acceptance Criteria

1. Encode the user session into an access token using Guardian.
2. Have the token time out after 1 day.

### PR DESCRIPTION
**Pull request by bterone**
Closes #21 

## What happened

✅: Add login API so we can log in via postman. 
✅: Create an authenticated API route for future routes that require authentication
✅: Took the pattern matching of routes and help with the proper configs from @hanam1ni's implementation 😄 
- [Router Plug](https://github.com/hanam1ni/google-crawler/blob/development/lib/google_crawler_web/plugs/router.ex)
- [Configs](https://github.com/hanam1ni/google-crawler/blob/development/config/config.exs#L33) Gave me too many headaches because Ueberauth didn't document about this anywhere that I could find 🤦 

## Insight

- Tried to have the API connect to another Google Client that is built for [limited-input devices](https://developers.google.com/identity/protocols/oauth2/limited-input-device), but the problem was that I wasn't sure how to config `ueberauth` to go to two clients depending on which route was called. 🤔

I was thinking of having two different Oauth libraries, but that seemed a bit odd, 😂 
I could look to refactor that in the future.

- When a user connects to an authenticated route, the flow is similar to the web, after verifying the token, we can set the user to the assigns and continue. Otherwise, it will render an `unauthorized` error.
 
#### `/api/auth/google`
Screen Shot 2021-06-21 at 10 07 11 AM

#### `Authenticated route test`
- With Token
Screen Shot 2021-06-21 at 10 11 48 AM

- Without Token
Screen Shot 2021-06-21 at 10 11 58 AM

### CODE REVIEW

File: config/prod.secret.exs [hunk 1]
```diff
@@ -35,6 +35,10 @@ config :ueberauth, Ueberauth.Strategy.Google.OAuth,
   client_id: System.get_env("GOOGLE_CLIENT_ID"),
   client_secret: System.get_env("GOOGLE_CLIENT_SECRET")
 
+config :g_searcher, GSearcher.Tokenizer,
+  issuer: "g_searcher",
+  secret_key: System.get_env("GUARDIAN_SECRET")
```
Comment (andyduong1920) [inline config/prod.secret.exs]: Small comment but we can change to `fetch_env!` here and other places in this file too :), just a recommend :)

File: lib/g_searcher/tokenizer.ex [hunk 2]
```diff
@@ -0,0 +1,21 @@
+defmodule GSearcher.Tokenizer do
+  use Guardian, otp_app: :g_searcher
+
+  alias GSearcher.Accounts
+
+  def generate_access_token(user) do
+    encode_and_sign(user, %{}, ttl: {1, :days}, token_type: "access")
```
Comment (andyduong1920) [inline lib/g_searcher/tokenizer.ex]: how about making it to be 1 line method 🤔 for here and the `subject_for_token` method too just a recommend :)

File: lib/g_searcher_web/controllers/api/authentication_controller.ex [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        conn
+        |> put_status(:ok)
+        |> render("token.json", %{access_token: access_token})
+
+      {:error, _reason} ->
+        ErrorHandler.render_error_json(conn, :bad_request, "Unable to sign in")
+    end
+  end
+
+  defp insert_or_update_user(user_params) do
```
Comment (andyduong1920) [inline lib/g_searcher_web/controllers/api/authentication_controller.ex]: do we have any update action within this block 🤔, I can see only `create_user` method 🤔

**Review by hanam1ni — APPROVED** [review]
LGTM 👍

**Review by andyduong1920 — APPROVED** [review]

File: lib/g_searcher_web/controllers/api/authentication_controller.ex (hunk 3 above)
Comment (bterone) [reply lib/g_searcher_web/controllers/api/authentication_controller.ex]: Nice catch 😄 , I've renamed this to `fetch_or_insert_user` instead d34228c5228008190d1ede8a642a82627222c07e.

**Pull request merged**